From 783c94f5611ff6d9d706564717acde4d47e5dea9 Mon Sep 17 00:00:00 2001
From: Pawan Gupta <pawan.kumarx.gupta@intel.com>
Date: Fri, 28 Oct 2016 16:00:06 -0700
Subject: [PATCH 512/639] mmc: wait for prev req complete before part switch

User partition switching was being done after every mmc request.
This is leading to mmc error:
mmc0: Timeout waiting for hardware interrupt.

This error is coming when there are pending mmc request for non-user
partition and we switch to user partition before a non-user partition
requests are complete.

Changed the user partition switch sequence to just before releasing the
host and hw mutex. At this point we are sure that there are no pending
atom mmc requests.

Signed-off-by: Pawan Gupta <pawan.kumarx.gupta@intel.com>
---
 drivers/mmc/card/block.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/mmc/card/block.c b/drivers/mmc/card/block.c
index f9ed35f..adca560 100644
--- a/drivers/mmc/card/block.c
+++ b/drivers/mmc/card/block.c
@@ -2022,17 +2022,18 @@ static int mmc_blk_issue_rq(struct mmc_queue *mq, struct request *req)
 	}
 
 out:
-	mmc_blk_part_switch_to_up(card);
-
 	if ((!req && !(mq->flags & MMC_QUEUE_NEW_REQUEST)) ||
-	     (cmd_flags & MMC_REQ_SPECIAL_MASK))
+	    (cmd_flags & MMC_REQ_SPECIAL_MASK)) {
 		/*
 		 * Release host when there are no more requests
 		 * and after special request(discard, flush) is done.
 		 * In case sepecial request, there is no reentry to
 		 * the 'mmc_blk_issue_rq' with 'mqrq_prev->req'.
 		 */
+		mmc_blk_part_switch_to_up(card);
 		mmc_put_card(card);
+	}
+
 	return ret;
 }
 
-- 
2.10.1

